### PR TITLE
[InstCombine] Fold Xor with or disjoint

### PR DESCRIPTION
**Pull request by AmrDeveloper**
Implement a missing optimization fold `(X | Y) ^ M to (X ^ M) ^ Y` and `(X | Y) ^ M to (Y ^ M) ^ X`

### CODE REVIEW

**Comment by DTeachs** [comment]
You can make this m_AddLike

See: https://alive2.llvm.org/ce/z/Fzma9H

**Comment by AmrDeveloper** [comment]
> You can make this m_AddLike
> 
> See: https://alive2.llvm.org/ce/z/Fzma9H

Cool, i will add it on another PR after land this

Thank you

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 1]
```diff
@@ -4056,6 +4056,25 @@ Instruction *InstCombinerImpl::visitOr(BinaryOperator &I) {
   return nullptr;
 }
 
+static Instruction *foldXorToOr(BinaryOperator &I,
+                                InstCombiner::BuilderTy &Builder) {
+  assert(I.getOpcode() == Instruction::Xor);
+  Value *A, *B, *C, *D, *E;
+
+  // ((select C, A, B) | E) ^ D) -> (select C, A ^ D, B ^ D) | E)
+  if (match(&I,
```
Comment (elhewaty) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:4065]: move this condition to the `InstCombinerImpl::visitXor` function.
I can help with the condition, but I let you think.
further hints:
I think I should refer to @dtcxzyw to double-check my suggestions:
- I think the issue was about matching constants in place of `A`, `B`, and `D`, so It's better to use `m_APInt` instead.

File: llvm/test/Transforms/InstCombine/disjoint_or.ll [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+; CHECK-NEXT:    [[TMP1:%.*]] = select i1 [[C]], <2 x i32> <i32 4, i32 4>, <2 x i32> zeroinitializer
+; CHECK-NEXT:    [[XOR:%.*]] = or disjoint <2 x i32> [[TMP1]], [[SHL]]
+; CHECK-NEXT:    ret <2 x i32> [[XOR]]
+;
+  %s = select i1 %c, <2 x i32> <i32 0, i32 0>, <2 x i32> <i32 4, i32 4>
+  %shl = shl <2 x i32> %a, <i32 4, i32 4>
+  %or = or disjoint <2 x i32> %s, %shl
+  %xor = xor <2 x i32> %or, <i32 4, i32 4>
+  ret <2 x i32> %xor
+}
```
[the remark this answers is not in the document]
Comment (elhewaty) [reply llvm/test/Transforms/InstCombine/disjoint_or.ll:32]: Also, we need test cases for commutative operations (`or` and `xor`) to see if the optimization works when the LHS and RHS are shuffled.
hmm, I think we need only for or as it has a non-const operand `A`, but we don't need for `xor` as the canonicalizer always places constants on the right-hand side.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp (hunk 1 above)
Comment (elhewaty) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:4065]: Will this check be profitable if the `or` has more than one use? 🤔

**Comment by AmrDeveloper** [comment]
> It's my fault for not providing generalized proof of the issue, but I don't think your pattern here is correct. IIUC, you try to fold `(A | B) ^ C` into `(A ^ C) | B` iff `A ^ C` can be simplified, but it's wrong: https://alive2.llvm.org/ce/z/s7VV8u The correct pattern should be `(A | B) ^ C --> (A ^ C) ^ B`: https://alive2.llvm.org/ce/z/u2vco3 The key is to treat `or disjoint` as `xor`.

Thank you, i change the pattern, test code ...etc

**Comment by elhewaty** [comment]
> The correct pattern should be (A | B) ^ C --> (A ^ C) ^ B: https://alive2.llvm.org/ce/z/u2vco3
The key is to treat or disjoint as xor.

Is this optimization profitable? We still have the same number of instructions, I think it depends on the target architecture if it has a special handling or if the pipelining prefers XOR.

this is just a guess.

**Comment by XChy** [comment]
> > The correct pattern should be (A | B) ^ C --> (A ^ C) ^ B: https://alive2.llvm.org/ce/z/u2vco3
> > The key is to treat or disjoint as xor.
> 
> Is this optimization profitable? We still have the same number of instructions, I think it depends on the target architecture if it has a special handling or if the pipelining prefers XOR.
> 
> this is just a guess.

This fold is profitable if `A ^ C` folds.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp (hunk 1 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:4065]: > Will this check be profitable if the `or` has more than one use? 🤔

Probably not.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 3]
```diff
@@ -4706,6 +4706,13 @@ Instruction *InstCombinerImpl::visitXor(BinaryOperator &I) {
     return Xor;
 
   Value *X, *Y;
+
+  // (A | B) ^ C -> (A ^ C) ^ B
+  if (match(Op0, m_OneUse(m_c_DisjointOr(m_Value(X), m_Value(Y))))) {
```
Comment (elhewaty) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:4711]: I don't think this will handle `C ^ (A | B)`!

``` suggestion
  if (match(&I, m_c_Xor(m_OneUse(m_c_DisjointOr(m_Value(X), m_Value(Y))), m_Value(M)))) {
 ```
  There's an `M` variable declared above if this is confusing you can declare a variable *Z in the `if` clause
  
  ```
   if (Value *Z; ...)
  ```

File: llvm/test/Transforms/InstCombine/xor.ll [hunk 4]
```diff
@@ -1502,8 +1502,8 @@ entry:
 define <2 x i32> @or_disjoint_with_xor_vec(<2 x i32> %a, < 2 x i32> %b, <2 x i32> %c) {
 ; CHECK-LABEL: @or_disjoint_with_xor_vec(
 ; CHECK-NEXT:  entry:
-; CHECK-NEXT:    [[OR:%.*]] = or disjoint <2 x i32> [[A:%.*]], [[B:%.*]]
-; CHECK-NEXT:    [[XOR:%.*]] = xor <2 x i32> [[OR]], [[C:%.*]]
+; CHECK-NEXT:    [[TMP0:%.*]] = xor <2 x i32> [[A:%.*]], [[C:%.*]]
+; CHECK-NEXT:    [[XOR:%.*]] = xor <2 x i32> [[TMP0]], [[B:%.*]]
 ; CHECK-NEXT:    ret <2 x i32> [[XOR]]
 ;
 entry:
```
Comment (elhewaty) [inline llvm/test/Transforms/InstCombine/xor.ll:1509]: Can you please add more negative test cases? 
you are using `m_OneUse` we should have a test to check if this works as expected.
and a test for the `C ^ (A | B)` case.

**Review by elhewaty — COMMENTED** [review]
this is a great work, thanks.

File: llvm/test/Transforms/InstCombine/xor.ll (hunk 4 above)
Comment (AmrDeveloper) [reply llvm/test/Transforms/InstCombine/xor.ll:1522]: Done

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp (hunk 3 above)
Comment (AmrDeveloper) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp]: Thanks for suggestion, i used M

**Comment by AmrDeveloper** [comment]
CI timeout of linux, green on windows !!

**Review by dtcxzyw — COMMENTED** [review]
Can you add some tests to demonstrate the usefulness of this patch (e.g., the original issue https://github.com/llvm/llvm-project/issues/79266)?

**Comment by AmrDeveloper** [comment]
> Can you add some tests to demonstrate the usefulness of this patch (e.g., the original issue #79266)?

Done

**Comment by AmrDeveloper** [comment]
> Can you add some tests to demonstrate the usefulness of this patch (e.g., the original issue #79266)?

Any idea why it timeout on `ninja runtimes C++03` on linux CI, i did rebase to the code but still same issue

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 5]
```diff
@@ -4706,6 +4706,15 @@ Instruction *InstCombinerImpl::visitXor(BinaryOperator &I) {
...
 
   Value *X, *Y;
+
+  // (A | B) ^ C -> (A ^ C) ^ B
+  // C ^ (A | B) -> B ^ (A ^ C)
+  if (match(&I, m_c_Xor(m_OneUse(m_c_DisjointOr(m_Value(X), m_Value(Y))),
+                        m_Value(M)))) {
+    Value *XorAC = Builder.CreateXor(X, M);
+    return BinaryOperator::CreateXor(XorAC, Y);
+  }
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:4716]: Is this desirable? I'd say `disjoint or` is preferable to an `xor` in most if not all respects.

I would limit this if `X ^ M` folds. You can test it with:
`if (Value * XorAC = simplifyBinOp(Instruction::Xor, X, M, SQ)`

**Comment by goldsteinn** [comment]
> > Can you add some tests to demonstrate the usefulness of this patch (e.g., the original issue #79266)?
> 
> Any idea why it timeout on `ninja runtimes C++03` on linux CI, i did rebase to the code but still same issue

Probably an inf loop. Typically we do `xor` -> `disjoint or`. If the three value all dont share bits I think this will inf loop between `(xor  (or disjoint x, y), z)` -> `(xor (xor x, z), y)`

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp (hunk 5 above)
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:4716]: You could also try if `Y, M` simplify.

Comment (AmrDeveloper) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:4716]: So can we do

if X ^ M folds or if Y ^ M folds -> (X ^ M) ^ Y

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 6]
```diff
@@ -101,16 +101,16 @@ Value *InstCombinerImpl::insertRangeTest(Value *V, const APInt &Lo,
...
+  AMask_AllOnes = 1,
+  AMask_NotAllOnes = 2,
+  BMask_AllOnes = 4,
+  BMask_NotAllOnes = 8,
+  Mask_AllZeros = 16,
+  Mask_NotAllZeros = 32,
+  AMask_Mixed = 64,
+  AMask_NotMixed = 128,
+  BMask_Mixed = 256,
+  BMask_NotMixed = 512
```
Comment (elhewaty) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:113]: This is unrelated.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 7]
```diff
@@ -172,15 +172,16 @@ static unsigned conjugateICmpMask(unsigned Mask) {
             << 1;
 
   NewMask |= (Mask & (AMask_NotAllOnes | BMask_NotAllOnes | Mask_NotAllZeros |
-                      AMask_NotMixed | BMask_NotMixed))
-             >> 1;
+                      AMask_NotMixed | BMask_NotMixed)) >>
+             1;
```
Comment (elhewaty) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:177]: this is too.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 8]
```diff
@@ -519,9 +520,9 @@ static Value *foldLogOpOfMaskedICmps(ICmpInst *LHS, ICmpInst *RHS, bool IsAnd,
   if (Mask == 0) {
     // Even if the two sides don't share a common pattern, check if folding can
     // still happen.
-    if (Value *V = foldLogOpOfMaskedICmpsAsymmetric(
-            LHS, RHS, IsAnd, A, B, C, D, E, PredL, PredR, LHSMask, RHSMask,
-            Builder))
+    if (Value *V = foldLogOpOfMaskedICmpsAsymmetric(LHS, RHS, IsAnd, A, B, C, D,
+                                                    E, PredL, PredR, LHSMask,
+                                                    RHSMask, Builder))
```
Comment (elhewaty) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:525]: Please fix the unrelated changes.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp (hunk 6 above)
Comment (AmrDeveloper) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:113]: Yes it from the formatter, i notice them but don't want to push and restart CI, once end i will update it and solve any other comments if there are new

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp (hunk 5 above)
Comment (AmrDeveloper) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp]: Now i checked simplify and update samples, this is alive2 for current test depending that A ^ M will folded

https://alive2.llvm.org/ce/z/NLhfhN

**Comment by AmrDeveloper** [comment]
> > > Can you add some tests to demonstrate the usefulness of this patch (e.g., the original issue #79266)?
> > 
> > Any idea why it timeout on `ninja runtimes C++03` on linux CI, i did rebase to the code but still same issue
> 
> Probably an inf loop. Typically we do `xor` -> `disjoint or`. If the three value all dont share bits I think this will inf loop between `(xor  (or disjoint x, y), z)` -> `(xor (xor x, z), y)`

Yes it passed now

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 9]
```diff
@@ -4693,7 +4693,21 @@ Instruction *InstCombinerImpl::visitXor(BinaryOperator &I) {
...
   // have already taken care of those cases.
   Value *Op0 = I.getOperand(0), *Op1 = I.getOperand(1);
-  Value *M;
+  Value *X, *Y, *M;
+
+  // (A | B) ^ C -> (A ^ C) ^ B
+  // C ^ (A | B) -> B ^ (A ^ C)
+  if (match(&I, m_c_Xor(m_OneUse(m_c_DisjointOr(m_Value(X), m_Value(Y))),
+                        m_Value(M)))) {
+    if (Value *XorAC = simplifyBinOp(Instruction::Xor, X, M, SQ)) {
```
Comment (XChy) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:4702]: Don’t use braces on single-statement body of if. See https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

Comment (AmrDeveloper) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp]: Thank you, Done

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp (hunk 5 above)
Comment (XChy) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp]: With `simplifyBinOp`, it could not handle the motivating case anymore. @dtcxzyw, could you test the impact of this patch? If it optimizes other cases still, I wouldn't argue about it.

Comment (XChy) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp]: An unrelated idea occurred to me: could we rank `select cond, C1, C2` with the same number as constants in the Reassociate pass?

**Comment by AmrDeveloper** [comment]
Any updates for this PR review @XChy, @goldsteinn, @dtcxzyw?

**Comment by goldsteinn** [comment]
LGTM

**Comment by AmrDeveloper** [comment]
@nikic Can you check this PR when you have time

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 10]
```diff
@@ -4693,7 +4693,19 @@ Instruction *InstCombinerImpl::visitXor(BinaryOperator &I) {
   // calls in there are unnecessary as SimplifyDemandedInstructionBits should
   // have already taken care of those cases.
   Value *Op0 = I.getOperand(0), *Op1 = I.getOperand(1);
-  Value *M;
+  Value *X, *Y, *M;
+
+  // (A | B) ^ C -> (A ^ C) ^ B
+  // C ^ (A | B) -> B ^ (A ^ C)
+  if (match(&I, m_c_Xor(m_OneUse(m_c_DisjointOr(m_Value(X), m_Value(Y))),
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:4700]:
```suggestion
  if (match(&I, m_c_Xor(m_OneUse(m_DisjointOr(m_Value(X), m_Value(Y))),
```
Commutative matcher doesn't make sense if you are only capturing values.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 11]
```diff
@@ -4693,7 +4693,19 @@ Instruction *InstCombinerImpl::visitXor(BinaryOperator &I) {
   // calls in there are unnecessary as SimplifyDemandedInstructionBits should
   // have already taken care of those cases.
   Value *Op0 = I.getOperand(0), *Op1 = I.getOperand(1);
-  Value *M;
+  Value *X, *Y, *M;
+
+  // (A | B) ^ C -> (A ^ C) ^ B
+  // C ^ (A | B) -> B ^ (A ^ C)
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:4699]:
```suggestion
  // (X | Y) ^ M -> (X ^ M) ^ Y
  // (X | Y) ^ M -> (Y ^ M) ^ X
```
Try to use the same names as in the code to make it easier to compare.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 12]
```diff
@@ -4693,7 +4693,19 @@ Instruction *InstCombinerImpl::visitXor(BinaryOperator &I) {
...
+  if (match(&I, m_c_Xor(m_OneUse(m_c_DisjointOr(m_Value(X), m_Value(Y))),
+                        m_Value(M)))) {
+    if (Value *XorAC = simplifyBinOp(Instruction::Xor, X, M, SQ))
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:4702]:
```suggestion
    if (Value *XorAC = simplifyBinOp(Instruction::Xor, X, M, SQ.getWithInstruction(&I)))
```

File: llvm/test/Transforms/InstCombine/xor.ll [hunk 13]
```diff
@@ -1485,3 +1485,171 @@ define i4 @PR96857_xor_without_noundef(i4  %val0, i4  %val1, i4 %val2) {
...
+  ret i32 %xor
+}
+
+define i32 @xor_with_or_disjoint(i32 %a, i32 %b, i32 %c) {
+; CHECK-LABEL: @xor_with_or_disjoint(
+; CHECK-NEXT:  entry:
+; CHECK-NEXT:    ret i32 [[B:%.*]]
+;
+entry:
+  %or = or disjoint i32 %a, %b
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/xor.ll:1506]: Should also have a commuted variant with `%b, %a`, if I didn't miss is somewhere.

File: llvm/test/Transforms/InstCombine/xor.ll [hunk 14]
```diff
@@ -1485,3 +1485,171 @@ define i4 @PR96857_xor_without_noundef(i4  %val0, i4  %val1, i4 %val2) {
...
+  %xor = xor i32 %or, %a
+  ret i32 %xor
+}
+
+define i32 @xor_with_or_disjoint(i32 %a, i32 %b, i32 %c) {
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/xor.ll:1500]: Unused argument `%c`.

**Review by nikic — COMMENTED** [review]
Looks reasonable.

A possible alternative would be to try to integrate this inside SimplifyAssociativeOrCommutative, by replacing the opcode `==` matches there with a stronger check that knows that `or disjoint` is compatible with or, xor and add. I'm okay with the separate fold though.

File: llvm/test/Transforms/InstCombine/xor.ll (hunk 14 above)
Comment (AmrDeveloper) [reply llvm/test/Transforms/InstCombine/xor.ll]: Done

File: llvm/test/Transforms/InstCombine/xor.ll (hunk 13 above)
Comment (AmrDeveloper) [reply llvm/test/Transforms/InstCombine/xor.ll:1506]: Done

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp (hunk 12 above)
Comment (AmrDeveloper) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp]: Done

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp (hunk 11 above)
Comment (AmrDeveloper) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp]: Done

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp (hunk 10 above)
Comment (AmrDeveloper) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp]: Done

**Comment by nikic** [comment]
> Fixes: #79266

Fixes -> "Related to" or something? I don't think this PR fixes that issue as-is.

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 15]
```diff
@@ -4693,7 +4693,20 @@ Instruction *InstCombinerImpl::visitXor(BinaryOperator &I) {
...
   Value *Op0 = I.getOperand(0), *Op1 = I.getOperand(1);
-  Value *M;
+  Value *X, *Y, *M;
+
+  // (X | Y) ^ M -> (X ^ M) ^ Y
+  // (X | Y) ^ M -> (Y ^ M) ^ X
+  if (match(&I, m_c_Xor(m_OneUse(m_DisjointOr(m_Value(X), m_Value(Y))),
+                        m_Value(M)))) {
+    if (Value *XorAC =
+            simplifyBinOp(Instruction::Xor, X, M, SQ.getWithInstruction(&I)))
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:4703]:
```suggestion
            simplifyXorInst(X, M, SQ.getWithInstruction(&I)))
```

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 16]
```diff
@@ -4693,7 +4693,20 @@ Instruction *InstCombinerImpl::visitXor(BinaryOperator &I) {
...
+      return BinaryOperator::CreateXor(XorAC, Y);
+
+    if (Value *XorBC = simplifyBinOp(Instruction::Xor, Y, M, SQ))
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:4706]:
```suggestion
    if (Value *XorBC = simplifyXorInst(Y, M, SQ.getWithInstruction(&I)))
```

**Review by nikic — APPROVED** [review]
LGTM

**Comment by AmrDeveloper** [comment]
> > Fixes: #79266
> 
> Fixes -> "Related to" or something? I don't think this PR fixes that issue as-is.

Yes we changed the case, i will remove issue mention and describe the current fix

Thank you
